README
tsdumper2
=========
tsdumper2 reads incoming mpeg transport stream over UDP/RTP and then
records it to disk. The files names are generated based on preconfigured
time interval.

License
=======
tsdumper2 is released under GNU GPL v2.

Development
===========
The development is done using git. tsdumper2 repository is hosted
at http://github.com/gfto/tsdumper2

To clone the repository issue the following commands:

   git clone git://github.com/gfto/tsdumper2.git
   cd tsdumper2
   git submodule init
   git submodule update
   make

The code is developed and tested under modern Linux. It is also
compiled from time to time under OS X but is not tested there.

To see all Makefile targets run 'make help'.

Updating the code
=================
To update cloned tsdumper2, go to the directory where the repository
is cloned and run the following commands:

   git fetch origin
   git merge origin/master
   git submodule update
   make clean all

tsdumper2's master branch should always be useful so it is safe to
use it instead of official release. The master branch will always
be better than any released version.

Command line parameters
=======================
tsdumper2 is controlled using command line parameters. For more information
about the parameters see the man page. Here is a list of supported command
line parameters:

        Usage: tsdumper2 -n <name> -i <input>

Settings:
 -n --prefix <name>         | Filename prefix.
 -s --seconds <seconds>     | How much to save (default: 60 sec).
 -d --output-dir <dir>      | Startup directory (default: .).
 -D --create-dirs           | Save files in subdirs YYYY/MM/DD/HH/file.

Input options:
 -i --input <source>        | Where to read from.
                            .  -i udp://224.0.0.1:5000    (v4 multicast)
                            .  -i udp://[ff01::1111]:5000 (v6 multicast)
                            .  -i rtp://224.0.0.1:5000    (v4 RTP input)
                            .  -i rtp://[ff01::1111]:5000 (v6 RTP input)
 -z --input-ignore-disc     | Do not report discontinuty errors in input.
 -4 --ipv4                  | Use only IPv4 addresses.
 -6 --ipv6                  | Use only IPv6 addresses.

Examples
========
To get a quick start here are some example command lines.

Examples:

   # Read multicast stream from 239.78.78.78:5000 and save it in
   # files with test- prefix
   tsdumper2 --input udp://239.78.78.78:5000/ --prefix test

   # Same as above but create directories YYYY/MM/DD/HH and put
   # files into the directory.
   tsdumper2 --input udp://239.78.78.78:5000/ --prefix test --create-dirs

   # Same as above but create directories YYYY/MM/DD/HH and put
   # files into the directory and create new file each 10 seconds.
   tsdumper2 --input udp://239.78.78.78:5000/ --prefix test --create-dirs --seconds 10

Reporting bugs
==============
If you think you have found bug in tsdumper2, please report it to the
e-mail listed in Contact section (see below) of this README file.

Releases
========
Official releases can be downloaded from tsdumper2 home page which is

   http://georgi.unixsol.org/programs/tsdumper2/

Contact
=======
For patches, bug reports, complaints and so on send e-mail to

   Georgi Chorbadzhiyski <georgi@unixsol.org>


ChangeLog
xxxx-xx-xx : Version -dev
 * Add support for RTP input.

2013-07-22 : Version 0.9
 * Initial public release.


udp.c
/*
 * UDP functions
 * Copyright (C) 2013 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License (COPYING file) for more details.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>

#include "tsdumper2.h"

#ifndef IPV6_ADD_MEMBERSHIP
#define IPV6_ADD_MEMBERSHIP IPV6_JOIN_GROUP
#define IPV6_DROP_MEMBERSHIP IPV6_LEAVE_GROUP
#endif

int ai_family = AF_UNSPEC;

static int is_multicast(struct sockaddr_storage *addr) {
	int ret = 0;
	switch (addr->ss_family) {
	case AF_INET: {
		struct sockaddr_in *addr4 = (struct sockaddr_in *)addr;
		ret = IN_MULTICAST(ntohl(addr4->sin_addr.s_addr));
		break;
	}
	case AF_INET6: {
		struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *)addr;
		ret = IN6_IS_ADDR_MULTICAST(&addr6->sin6_addr);
		break;
	} }
	return ret;
}

static int bind_addr(const char *hostname, const char *service, int socktype, struct sockaddr_storage *addr, int *addrlen, int *sock) {
	struct addrinfo hints, *res, *ressave;
	int n, ret = -1;

	memset(&hints, 0, sizeof(struct addrinfo));
	hints.ai_family = ai_family;
	hints.ai_socktype = socktype;

	n = getaddrinfo(hostname, service, &hints, &res);
	if (n < 0) {
		p_info("ERROR: getaddrinfo(%s): %s\n", hostname, gai_strerror(n));
		return ret;
	}

	ressave = res;
	while (res) {
		*sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
		if (*sock > -1) {
			int on = 1;
			setsockopt(*sock, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
			set_sock_nonblock(*sock);
			if (bind(*sock, res->ai_addr, res->ai_addrlen) == 0) {
				memcpy(addr, res->ai_addr, res->ai_addrlen);
				*addrlen = res->ai_addrlen;
				ret = 0;
				goto OUT;
			} else {
				char str_addr[INET6_ADDRSTRLEN];
				my_inet_ntop(res->ai_family, res->ai_addr, str_addr, sizeof(str_addr));
				p_err("bind: %s:%s (%s)", hostname, service, str_addr);
			}
			close(*sock);
			*sock = -1;
		}
		res = res->ai_next;
	}
OUT:
	freeaddrinfo(ressave);

	return ret;
}

static int join_multicast_group(int sock, struct sockaddr_storage *addr) {
	switch (addr->ss_family) {
	case AF_INET: {
		struct ip_mreq mreq;
		mreq.imr_multiaddr.s_addr = ((struct sockaddr_in *)addr)->sin_addr.s_addr;
		mreq.imr_interface.s_addr = INADDR_ANY;

		if (setsockopt(sock, IPPROTO_IP, IP_ADD_MEMBERSHIP, (const void *)&mreq, sizeof(mreq)) < 0) {
			p_err("setsockopt(IP_ADD_MEMBERSHIP)");
			return -1;
		}
		break;
	}

	case AF_INET6: {
		struct ipv6_mreq mreq6;
		memcpy(&mreq6.ipv6mr_multiaddr, &(((struct sockaddr_in6 *)addr)->sin6_addr), sizeof(struct in6_addr));
		mreq6.ipv6mr_interface = 0; // interface index, will be set later

		if (setsockopt(sock, IPPROTO_IPV6, IPV6_ADD_MEMBERSHIP, &mreq6, sizeof(mreq6)) < 0) {
			p_err("setsockopt(IPV6_ADD_MEMBERSHIP)");
			return -1;
		}
		break;
	}
	}

	return 0;
}

int udp_connect_input(struct io *io) {
	struct sockaddr_storage addr;
	int addrlen = sizeof(addr);
	int sock = -1;

	memset(&addr, 0, sizeof(addr));

	p_info("Connecting input to %s port %s\n", io->hostname, io->service);
	if (bind_addr(io->hostname, io->service, SOCK_DGRAM, &addr, &addrlen, &sock) < 0)
		return -1;

	/* Set receive buffer size to ~4.0MB */
	int bufsize = (4000000 / 1316) * 1316;
	setsockopt(sock, SOL_SOCKET, SO_RCVBUF, (void *)&bufsize, sizeof(bufsize));

	if (is_multicast(&addr)) {
		if (join_multicast_group(sock, &addr) < 0) {
			close(sock);
			return -1;
		}
	}

	io->fd = sock;
	p_info("Input connected to fd:%d\n", io->fd);

	return 1;
}


util.c
/*
 * Utility functions
 * Copyright (C) 2013 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License (COPYING file) for more details.
 *
 */

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <time.h>
#include <string.h>
#include <errno.h>
#include <arpa/inet.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "tsdumper2.h"

#ifdef __linux__
#include <sys/prctl.h>

void set_thread_name(char *thread_name) {
	prctl(PR_SET_NAME, thread_name, NULL, NULL, NULL);
}

#else
void set_thread_name(char *thread_name) {
    (void)thread_name;
}

#endif

int parse_host_and_port(char *input, struct io *io) {
	int port_set = 0;
	char *p, *proto;
	io->type = UDP;
	if (strlen(input) == 0)
		return 0;
	if (strstr(input, "udp://") == input)       io->type = UDP;
	else if (strstr(input, "rtp://") == input)  io->type = RTP;
	else
		die("Unsupported protocol (patch welcome): %s", input);
	proto = strstr(input, "://");
	if (proto)
		input = proto + 3;
	io->hostname = input;
	if (input[0] == '[') { // Detect IPv6 static address
		p = strrchr(input, ']');
		if (!p)
			die("Invalid IPv6 address format: %s\n", input);
		io->hostname = input + 1; // Remove first [
		*p = 0x00; // Remove last ]
		char *p2 = strchr(p + 1, ':');
		if (p2) {
			*p2 = 0x00;
			io->service = p2 + 1;
			port_set = 1;
		}
	} else {
		p = strrchr(input, ':');
		if (p) {
			*p = 0x00;
			io->service = p + 1;
			port_set = 1;
		}
	}
	if (io->service) {
		char *path = strstr(io->service, "/");
		if (path)
			path[0] = 0;
	}
	if (!port_set)
		die("Port is not set in the input url.");
	return 1;
}

char *my_inet_ntop(int family, struct sockaddr *addr, char *dest, int dest_len) {
	struct sockaddr_in  *addr_v4 = (struct sockaddr_in  *)addr;
	struct sockaddr_in6 *addr_v6 = (struct sockaddr_in6 *)addr;
	switch (family) {
		case AF_INET:
			return (char *)inet_ntop(AF_INET, &addr_v4->sin_addr, dest, dest_len);
			break;
		case AF_INET6:
			return (char *)inet_ntop(AF_INET6, &addr_v6->sin6_addr, dest, dest_len);
			break;
		default:
			memset(dest, 0, dest_len);
			strcpy(dest, "unknown");
			return dest;
	}
}

int create_dir(const char *dir, mode_t mode) {
	int ret = 0;
	unsigned int i;

	// Shortcut
	if (strchr(dir, '/') == NULL)
		return mkdir(dir, mode);

	char *d = strdup(dir);
	unsigned int dlen = strlen(dir);

	// Skip first char (it can be /)
	for (i = 1; i < dlen; i++) {
		if (d[i] != '/')
			continue;
		d[i] = '\0';
		ret = mkdir(d, mode);
		d[i] = '/';
		if (ret < 0 && errno != EEXIST)
			goto OUT;
	}
	ret = mkdir(d, mode);
OUT:
	free(d);
	return ret;
}

void die(const char *fmt, ...) {
	va_list args;
	va_start(args, fmt);
	fprintf(stderr, "ERROR: ");
	vfprintf(stderr, fmt, args);
	if (fmt[strlen(fmt) - 1] != '\n')
		fprintf(stderr, "\n");
	va_end(args);
	exit(EXIT_FAILURE);
}

void p_err(const char *fmt, ...) {
	va_list args;
	va_start(args, fmt);
	fprintf(stderr, "ERROR: ");
	vfprintf(stderr, fmt, args);
	if (fmt[strlen(fmt) - 1] != '\n')
		fprintf(stderr, "\n");
	va_end(args);
}

void p_info(const char *fmt, ...) {
	if (DEBUG > 0) {
		struct timeval tv;
		gettimeofday(&tv, NULL);
		fprintf(stdout, "%08ld.%08ld ", tv.tv_sec, tv.tv_usec);

		char date[64];
		struct tm tm;
		localtime_r(&tv.tv_sec, &tm);
		strftime(date, sizeof(date), "%F %H:%M:%S", localtime_r(&tv.tv_sec, &tm));
		fprintf(stdout, "%s | ", date);
	}
	va_list args;
	va_start(args, fmt);
	vfprintf(stdout, fmt, args);
	if (fmt[strlen(fmt) - 1] != '\n')
		fprintf(stdout, "\n");
	va_end(args);
	fflush(stdout);
}


util.h
/*
 * Utility functions header
 * Copyright (C) 2013 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License (COPYING file) for more details.
 *
 */
#ifndef UTIL_H
#define UTIL_H

#include <inttypes.h>
#include <arpa/inet.h>

void set_thread_name(char *thread_name);

int parse_host_and_port(char *input, struct io *io);
char *my_inet_ntop(int family, struct sockaddr *addr, char *dest, int dest_len);

int create_dir(const char *dir, mode_t mode);

#define p_dbg1(fmt, ...) \
	do { if (DEBUG > 0) p_info(fmt, __VA_ARGS__); } while(0)

#define p_dbg2(fmt, ...) \
	do { if (DEBUG > 1) p_info(fmt, __VA_ARGS__); } while(0)

__attribute__ ((format(printf, 1, 2))) void die(const char *fmt, ...);

__attribute__ ((format(printf, 1, 2))) void p_err(const char *fmt, ...);
__attribute__ ((format(printf, 1, 2))) void p_info(const char *fmt, ...);

#endif


process.c
/*
 * Process incoming data and save it into files.
 * Copyright (C) 2013 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License (COPYING file) for more details.
 *
 */
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>

#include "tsdumper2.h"

#define NO_UNLINK  0
#define UNLINK_OLD 1

#define ALIGN_DOWN(__src, __value) (__src - (__src % __value))

static mode_t dir_perm;

static int file_exists(char *filename) {
	return access(filename, W_OK) == 0;
}

static void format_output_filename(struct ts *ts, time_t file_time) {
	struct tm file_tm;
	localtime_r(&file_time, &file_tm);

	ts->output_startts = file_time;

	ts->output_filename[0] = '\0';
	strcat(ts->output_filename, ts->prefix);
	strcat(ts->output_filename, "-");
	strftime(ts->output_filename + strlen(ts->output_filename), OUTFILE_NAME_MAX, "%Y%m%d_%H%M%S-%s.ts", &file_tm);

	ts->output_dirname[0] = '\0';
	strftime(ts->output_dirname, OUTFILE_NAME_MAX, "%Y/%m/%d/%H", &file_tm);

	ts->output_full_filename[0] = '\0';
	snprintf(ts->output_full_filename, sizeof(ts->output_full_filename), "%s/%s",
		ts->output_dirname, ts->output_filename);
}

static void report_file_creation(struct ts *ts, char *text_prefix, char *filename) {
	char qdepth[32];
	qdepth[0] = '\0';
	if (ts->packet_queue->items)
		snprintf(qdepth, sizeof(qdepth), " (depth:%d)", ts->packet_queue->items);
	p_info("%s%s%s\n", text_prefix, filename, qdepth);
}

static void create_output_directory(struct ts *ts) {
	if (!ts->create_dirs)
		return;
	if (!file_exists(ts->output_dirname)) {
		p_info(" = Create directory %s", ts->output_dirname);
		create_dir(ts->output_dirname, dir_perm);
	}
}

static int create_output_file(struct ts *ts) {
	char *filename = ts->output_filename;
	if (ts->create_dirs)
		filename = ts->output_full_filename;
	create_output_directory(ts);
	report_file_creation(ts, " = Create new file ", filename);
	int fd = open(ts->output_filename, O_CREAT | O_WRONLY | O_TRUNC, 0644);
	if (fd < 0) {
		p_err("Can't create output file %s", ts->output_filename);
		return -1;
	}
	if (ts->create_dirs) {
		link(ts->output_filename, ts->output_full_filename);
	}
	return fd;
}

static int append_output_file(struct ts *ts) {
	char *filename = ts->output_filename;
	if (ts->create_dirs)
		filename = ts->output_full_filename;
	create_output_directory(ts);
	report_file_creation(ts, " + Append to file ", filename);
	int fd = open(ts->output_filename, O_APPEND | O_WRONLY);
	if (fd < 0) {
		p_err("Can't append to output file %s", ts->output_filename);
		return -1;
	}
	if (ts->create_dirs) {
		link(ts->output_filename, ts->output_full_filename);
	}
	return fd;
}

static void close_output_file(struct ts *ts, int unlink_file) {
	if (ts->output_fd > -1) {
		close(ts->output_fd);
		if (unlink_file && ts->create_dirs) {
			// The file is hard linked into the subdirectory. There is no need
			// to keep it in the main directory.
			unlink(ts->output_filename);
		}
	}
}

static void handle_files(struct ts *ts, struct packet *packet) {
	int file_time = ALIGN_DOWN(packet->ts.tv_sec, ts->rotate_secs);

	// Is this file already created?
	if (file_time <= ts->output_startts)
		return;

	close_output_file(ts, UNLINK_OLD);
	format_output_filename(ts, file_time);

	/*
	 * When tsdumper2 is started, try to continue writing into "current" file.
	 * This allows the program to be killed/restarted.
	 *
	 * If current file does not exist, create new file with the time of the start
	 * (not aligned to rotate_secs).
	 */
	int append = 0;
	if (ts->output_fd < 0) { // First file (or error).
		append = file_exists(ts->output_filename);
		if (!append) { // Create first file *NOT ALIGNED*
			format_output_filename(ts, packet->ts.tv_sec);
		}
	}

	ts->output_fd = append ? append_output_file(ts) : create_output_file(ts);
}

void *write_thread(void *_ts) {
	struct ts *ts = _ts;
	struct packet *packet;

	mode_t umask_val = umask(0);
	dir_perm = (0777 & ~umask_val) | (S_IWUSR | S_IXUSR);

	set_thread_name("tsdump-write");
	while ((packet = queue_get(ts->packet_queue))) {
		if (!packet->data_len)
			continue;

		time_t packet_time = packet->ts.tv_sec;
		time_t file_time   = ALIGN_DOWN(packet_time, ts->rotate_secs);
		p_dbg1(" - Got packet %d, size: %u, file_time:%lu packet_time:%lu depth:%d\n",
			packet->num, packet->data_len, file_time, packet_time, ts->packet_queue->items);

		handle_files(ts, packet);

		if (ts->output_fd > -1) {
			p_dbg2(" - Writing into fd:%d size:%d file:%s\n", ts->output_fd, packet->data_len, ts->output_filename);
			ssize_t written = write(ts->output_fd, packet->data, packet->data_len);
			if (written != packet->data_len) {
				p_err("Can not write data (fd:%d written %zd of %d file:%s)",
					ts->output_fd, written, packet->data_len, ts->output_filename);
			}
		}
		free_packet(packet);
	}
	close_output_file(ts, NO_UNLINK);
	return NULL;
}

static struct packet *add_to_queue(struct ts *ts) {
	queue_add(ts->packet_queue, ts->current_packet);
	ts->current_packet = alloc_packet(ts);
	return ts->current_packet;
}

void process_packets(struct ts *ts, uint8_t *ts_packet, ssize_t readen) {
	struct timeval now;
	struct packet *packet = ts->current_packet;

	if (packet->data_len + readen < PACKET_MAX_LENGTH) {
		// Add data to buffer
		memcpy(packet->data + packet->data_len, ts_packet, readen);
		packet->data_len += readen;
	} else {
		// Too much data, add to queue
		p_dbg1("*** Reached buffer end (%zd + %zd > %d)\n", packet->data_len + readen, readen, PACKET_MAX_LENGTH);
		packet = add_to_queue(ts);
	}

	if (!packet->ts.tv_sec)
		gettimeofday(&packet->ts, NULL);

	gettimeofday(&now, NULL);
	unsigned long long diff = timeval_diff_msec(&packet->ts, &now);
	if (diff > PACKET_MAX_TIME) {
		// Too much time have passed, add to queue
		p_dbg1("+++ Reached time limit (%llu > %d)\n", diff, PACKET_MAX_TIME);
		add_to_queue(ts);
	}
}


tsdumper2.h
/*
 * Data definitions
 * Copyright (C) 2013 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License (COPYING file) for more details.
 *
 */
#ifndef DATA_H
#define DATA_H

#include <pthread.h>
#include <inttypes.h>
#include <stdbool.h>

#include "libfuncs/libfuncs.h"

// Supported values 0, 1 and 2. Higher value equals more spam in the log.
#define DEBUG 0

// 7 * 188
#define FRAME_SIZE 1316

// 64k should be enough for everybody
#define THREAD_STACK_SIZE (64 * 1024)

enum io_type {
	UDP,
	RTP,
};

// 1.2MB = ~13Mbit/s
#define PACKET_MAX_LENGTH (1316 * 1024)

// Maximum packet fill time in ms
#define PACKET_MAX_TIME 1000

#define PREFIX_MAX_LENGTH 64

// PREFIX-20130717_000900-1374008940.ts (PREFIX-YYYYMMDD_HHMMSS-0123456789.ts)
#define OUTFILE_NAME_MAX  (PREFIX_MAX_LENGTH + 128)

#define NUM_PACKETS 16

struct packet {
	int					num;
	struct timeval		ts;							// packet start time
	int					allocated;					// set to true if the struct is dynamically allocated
	int					in_use;						// this packet is currently being used
	int					data_len;					// data length
	uint8_t				data[PACKET_MAX_LENGTH];	// the data
};

struct io {
	int					fd;
	enum io_type		type;
	char				*hostname;
	char				*service;
};

struct ts {
	char				*prefix;
	char				*output_dir;
	int					create_dirs;
	int					rotate_secs;
	int					ts_discont;
	struct io			input;

	pthread_attr_t		thread_attr;
	pthread_t			write_thread;

	struct packet		packets[NUM_PACKETS];
	int					packet_ptr;

	struct packet		*current_packet;
	QUEUE				*packet_queue;

	int					output_fd;
	time_t				output_startts;
	char				output_dirname[OUTFILE_NAME_MAX];
	char				output_filename[OUTFILE_NAME_MAX];
	char				output_full_filename[OUTFILE_NAME_MAX];
};

#include "util.h"

// From tsdumper2.c
struct packet *alloc_packet(struct ts *ts);
void free_packet(struct packet *packet);

// From process.c
void *write_thread(void *_ts);
void process_packets(struct ts *ts, uint8_t *ts_packet, ssize_t readen);

// From udp.c
int udp_connect_input(struct io *io);

#endif


tsdumper2.c
/*
 * tsdumper2
 * Copyright (C) 2013 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License (COPYING file) for more details.
 *
 */
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/resource.h>

#include "tsdumper2.h"

extern int ai_family;

#define PROGRAM_NAME "tsdumper2"
static const char *program_id = PROGRAM_NAME " v" VERSION " (git-" GIT_VER ", build date " BUILD_ID ")";

static int keep_running = 1;
static unsigned long long total_read;

static const char short_options[] = "n:s:d:i:z46DhV";

static const struct option long_options[] = {
	{ "prefix",				required_argument, NULL, 'n' },
	{ "seconds",			required_argument, NULL, 's' },
	{ "output-dir",			required_argument, NULL, 'd' },
	{ "create-dirs",		no_argument,       NULL, 'D' },

	{ "input",				required_argument, NULL, 'i' },
	{ "input-ignore-disc",	no_argument,       NULL, 'z' },
	{ "ipv4",				no_argument,       NULL, '4' },
	{ "ipv6",				no_argument,       NULL, '6' },

	{ "help",				no_argument,       NULL, 'h' },
	{ "version",			no_argument,       NULL, 'V' },

	{ 0, 0, 0, 0 }
};

static void show_help(struct ts *ts) {
	printf("%s\n", program_id);
	printf("Copyright (C) 2013 Unix Solutions Ltd.\n");
	printf("\n");
	printf("	Usage: " PROGRAM_NAME " -n <name> -i <input>\n");
	printf("\n");
	printf("Settings:\n");
	printf(" -n --prefix <name>         | Filename prefix.\n");
	printf(" -s --seconds <seconds>     | How much to save (default: %u sec).\n", ts->rotate_secs);
	printf(" -d --output-dir <dir>      | Startup directory (default: %s).\n", ts->output_dir);
	printf(" -D --create-dirs           | Save files in subdirs YYYY/MM/DD/HH/file.\n");
	printf("\n");
	printf("Input options:\n");
	printf(" -i --input <source>        | Where to read from.\n");
	printf("                            .  -i udp://224.0.0.1:5000    (v4 multicast)\n");
	printf("                            .  -i udp://[ff01::1111]:5000 (v6 multicast)\n");
	printf("                            .  -i rtp://224.0.0.1:5000    (v4 RTP input)\n");
	printf("                            .  -i rtp://[ff01::1111]:5000 (v6 RTP input)\n");
	printf(" -z --input-ignore-disc     | Do not report discontinuty errors in input.\n");
	printf(" -4 --ipv4                  | Use only IPv4 addresses.\n");
	printf(" -6 --ipv6                  | Use only IPv6 addresses.\n");
	printf("\n");
	printf("Misc options:\n");
	printf(" -h --help                  | Show help screen.\n");
	printf(" -V --version               | Show program version.\n");
	printf("\n");
}

extern char *optarg;
extern int optind, opterr, optopt;

static void parse_options(struct ts *ts, int argc, char **argv) {
	int j, input_addr_err = 1;
	opterr = 0; // Prevent printing of error messages for unknown options in getopt()
	while ((j = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
		if (j == '?')
			die("Unknown parameter '%s'.\n", argv[optind - 1]);
		switch (j) {
			case 'n': // --prefix
				ts->prefix = optarg;
				if (strlen(optarg) >= PREFIX_MAX_LENGTH)
					die("Prefix is longer than %d characters!", PREFIX_MAX_LENGTH);
				break;
			case 's': // --seconds
				ts->rotate_secs = atoi(optarg);
				break;
			case 'd': // --output-dir
				ts->output_dir = optarg;
				break;
			case 'D': // --create-dirs
				ts->create_dirs = !ts->create_dirs;
				break;
			case 'i': // --input
				input_addr_err = !parse_host_and_port(optarg, &ts->input);
				break;
			case 'z': // --input-ignore-disc
				ts->ts_discont = !ts->ts_discont;
				break;
			case '4': // --ipv4
				ai_family = AF_INET;
				break;
			case '6': // --ipv6
				ai_family = AF_INET6;
				break;
			case 'h': // --help
				show_help(ts);
				exit(EXIT_SUCCESS);
			case 'V': // --version
				printf("%s\n", program_id);
				exit(EXIT_SUCCESS);
		}
	}
	if (input_addr_err || !ts->prefix) {
		show_help(ts);
		if (!ts->prefix)
			fprintf(stderr, "ERROR: File name prefix is not set (--prefix XXX | -n XXX).\n");
		if (input_addr_err)
			fprintf(stderr, "ERROR: Input address is invalid (--input XXX | -i XXX).\n");
		exit(EXIT_FAILURE);
	}

	p_info("Prefix     : %s\n", ts->prefix);
	p_info("Input addr : %s://%s:%s/\n",
		ts->input.type == UDP ? "udp" :
		ts->input.type == RTP ? "rtp" : "???",
		ts->input.hostname, ts->input.service);
	p_info("Seconds    : %u\n", ts->rotate_secs);
	p_info("Output dir : %s (create directories: %s)\n", ts->output_dir,
		ts->create_dirs ? "YES" : "no");
	if (chdir(ts->output_dir) < 0)
		die("Can not change directory to %s: %s\n", ts->output_dir, strerror(errno));
}

void signal_quit(int sig) {
	if (!keep_running)
		raise(sig);
	keep_running = 0;
	p_info("Killed %s with signal %d\n", program_id, sig);
	signal(sig, SIG_DFL);
}

static void clear_packet(struct packet *p) {
	p->ts.tv_sec  = 0;
	p->ts.tv_usec = 0;
	p->data_len   = 0;
	p->allocated  = 0;
	p->in_use     = 0;
}

struct packet *alloc_packet(struct ts *ts) {
	// check for free static allocations
	struct packet *p;
	int i;
	for (i = 0; i < NUM_PACKETS; i++) {
		p = &ts->packets[i];
		if (!p->in_use) {
			p->in_use = 1;
			p_dbg2("STATIC packet, num %d\n", p->num);
			goto OUT;
		}
	}
	// Dynamically allocate packet
	p = malloc(sizeof(struct packet));
	if (!p)
		die("Can't alloc %lu bytes.\n", (unsigned long)sizeof(struct packet));
	clear_packet(p);
	p->num       = time(NULL);
	p->allocated = 1;
	p_dbg2("ALLOC  packet, num %d\n", p->num);
OUT:
	return p;
}

void free_packet(struct packet *packet) {
	if (!packet->allocated) {
		clear_packet(packet);
		return;
	}
	p_dbg2("FREE   packet, num %d\n", packet->num);
	free(packet);
}

#define RTP_HDR_SZ  12

static uint8_t ts_packet[FRAME_SIZE + RTP_HDR_SZ];
static uint8_t rtp_hdr[2][RTP_HDR_SZ];
static struct ts ts;

int main(int argc, char **argv) {
	int i;
	int have_data = 1;
	int ntimeouts = 0;
	int rtp_hdr_pos = 0, num_packets = 0;
	struct rlimit rl;

	if (getrlimit(RLIMIT_STACK, &rl) == 0) {
		if (rl.rlim_cur > THREAD_STACK_SIZE) {
			rl.rlim_cur = THREAD_STACK_SIZE;
			setrlimit(RLIMIT_STACK, &rl);
		}
	}

	memset(rtp_hdr[0], 0, RTP_HDR_SZ);
	memset(rtp_hdr[1], 0, RTP_HDR_SZ);

	for (i = 0; i < NUM_PACKETS; i++) {
		struct packet *p = &ts.packets[i];
		p->num = i + 1;
	}

	ts.ts_discont     = 1;
	ts.output_dir     = ".";
	ts.rotate_secs    = 60;
	ts.current_packet = alloc_packet(&ts);
	ts.output_fd      = -1;

	pthread_attr_init(&ts.thread_attr);
	size_t stack_size;
	pthread_attr_getstacksize(&ts.thread_attr, &stack_size);
	if (stack_size > THREAD_STACK_SIZE)
		pthread_attr_setstacksize(&ts.thread_attr, THREAD_STACK_SIZE);

	parse_options(&ts, argc, argv);

	ts.packet_queue   = queue_new();

	p_info("Start %s\n", program_id);

	switch (ts.input.type) {
	case UDP:
	case RTP:
		if (udp_connect_input(&ts.input) < 1)
			exit(EXIT_FAILURE);
		break;
	}

	signal(SIGCHLD, SIG_IGN);
	signal(SIGPIPE, SIG_IGN);

	signal(SIGINT , signal_quit);
	signal(SIGTERM, signal_quit);

	pthread_create(&ts.write_thread , &ts.thread_attr, &write_thread , &ts);

	int data_received = 0;
	do {
		ssize_t readen = -1;
		set_log_io_errors(0);
		switch (ts.input.type) {
		case UDP:
			readen = fdread_ex(ts.input.fd, (char *)ts_packet, FRAME_SIZE, 250, 4, 1);
			break;
		case RTP:
			readen = fdread_ex(ts.input.fd, (char *)ts_packet, FRAME_SIZE + RTP_HDR_SZ, 250, 4, 1);
			if (readen > RTP_HDR_SZ) {
				memcpy(rtp_hdr[rtp_hdr_pos], ts_packet, RTP_HDR_SZ);
				memmove(ts_packet, ts_packet + RTP_HDR_SZ, FRAME_SIZE);
				readen -= RTP_HDR_SZ;
				uint16_t ssrc  = (rtp_hdr[rtp_hdr_pos][2] << 8) | rtp_hdr[rtp_hdr_pos][3];
				uint16_t pssrc = (rtp_hdr[!rtp_hdr_pos][2] << 8) | rtp_hdr[!rtp_hdr_pos][3];
				rtp_hdr_pos = !rtp_hdr_pos;
				if (pssrc + 1 != ssrc && (ssrc != 0 && pssrc != 0xffff) && num_packets > 2)
					if (ts.ts_discont)
						p_info(" *** RTP discontinuity last_ssrc %5d, curr_ssrc %5d, lost %d packet ***\n",
							pssrc, ssrc, ((ssrc - pssrc)-1) & 0xffff);
				num_packets++;
			}
			break;
		}
		set_log_io_errors(1);
		if (readen < 0) {
			p_info(" *** Input read timeout ***\n");
			data_received = 0;
			ntimeouts++;
		} else {
			if (ntimeouts && readen > 0) {
				ntimeouts = 0;
			}
		}
		if (readen > 0) {
			if (!data_received) {
				p_info("Data received.\n");
				data_received = 1;
			}
			total_read += readen;
			process_packets(&ts, ts_packet, readen);
		}
		if (!keep_running)
			break;
	} while (have_data);

	queue_add(ts.packet_queue, ts.current_packet);
	queue_add(ts.packet_queue, NULL); // Exit write_thread
	pthread_join(ts.write_thread, NULL);

	p_info("Stop %s (bytes_processed:%llu).\n", program_id, total_read);

	queue_free(&ts.packet_queue);

	pthread_attr_destroy(&ts.thread_attr);

	exit(EXIT_SUCCESS);
}
